### ISSUE TITLE
Command-List

### ISSUE THREAD

**Issue by florianb** [florianb/atom-javaenv#1]
Which commands/services should be provided by the javaenv-package?

I envision a pluggable API for specific file-formats, so which commands/services would a Java-dependent-package expect?

```
javaenv = watchProject(project)
# should keep track of the project, provides all java-information by the javaenv-object

javaenv.dispose()
# disable tracking?

javaenv.classpaths()
# all cumulated classpaths of the project

javaenv.sourcefiles()
# all sourcefiles in the project

javaenv.temporaryArgs( options: {classpaths: true, sourcefiles: true} )
# path to a curated temporary @argsfile, containing optional content

javaenv.jvm-version?
javaenv.jvm-executable? ( path?)

javaenv.tools # list of tools as [tool-objects]
javaenv.tool('maven') # tool-object of maven
javaenv.tool('maven').classpaths() # get classpaths?
```

We should consider using services for inter-package-communication and expose only special curated commands to the user via the command-interface of Atom.

**Comment by keskiju** [comment]
I think that this plugin should be a java class loader and class dictionary that provides services for loading and analyzing classes found in classpath, analyzing and searching java source files on project path, commands for refreshing that data, etc. Since most of this is already implemented in autocomplete-java package and autocomplete-java requirements are much more demanding than requirements of linter-javac, I would like to take main responsibility for this.

**Comment by florianb** [comment]
That's okay - what do you think of?

**Comment by keskiju** [comment]
Basically it parses classes from classpath automatically (also watches class file changes) and builds a class dictionary for all java plugins to use. Perhaps the same for java source files also, with some javadoc capabilities. Most of this is already implemented. I will start planning these services once I have time to refactor autocomplete-java.

For simple requirements of linter-javac, I would suggest just a simple method:

const param = await javaenv.readBuildProperties(currentFilePath);

This would return all build properties that linter-javac needs: classpath, compile target directory, source file encoding, argsfile path, JAVA_HOME, etc. Build properties should be determined by current file path, because using cumulative classpath would lead to poor performance and wrong results.

**Comment by keskiju** [comment]
javaenv caches all build properties and refresh/clean command clears all caches. The readBuildProperties method loads properties from disk only in case of a cache miss.

**Comment by keskiju** [comment]
readBuildProperties reads properties both from disk (multple file formats supported) and from javaenv package settings.

**Comment by florianb** [comment]
Your ideas are good, i think the responsibility to assemble the final build-properties should not belong to the javaenv-package. I would appreciate if the javaenv-package could be something like a property-repository, giving easy access to needed values.

**Comment by florianb** [comment]
And i would like to try, if we could make use of third-party tools, like f.e. Maven or Eclipse-classpath-files. I imagine the possibility to decide Atom-javac picking up the project-configuration which is f.e. used for the ci-process.

**Comment by keskiju** [comment]
Yes, javaenv is a property repository. readBuildProperties returns properties as a Map. If you need many properties, you would not fetch them from javaenv one property at a time.

**Comment by florianb** [comment]
You're right.. :+1:

**Comment by keskiju** [comment]
Yes, readBuildProperties would eventually support multiple file formats (simple properties file, Eclipse .classpath, maven, ...) and determines the format automatically.

**Comment by florianb** [comment]
That sounds nice - to begin the implementation we could define an "elder"-interface which rawly consists out of your current functions.
As a second step we could draft the "new" API as a wrapper to the elder-functionality. Then we would have fastly the javaenv up and running and could discuss design-decisions more practically.

**Comment by keskiju** [comment]
Under the hood it would scan all class files, jar files and source files according to defined build properties and it would also watch for file changes. Using this info it can create class descriptions and build an index for searching. Usage would mostly be searching class and class member descriptions against index:

clazz = await index.findCurrentClass();
clazz = await index.findClass('com.domain.myapp.test.TestClass');
classes = await index.findClasses('TestC_');
classes = await index.findClassesBySimpleName('TestClass');
members = await index.findClassMembers('com.domain.myapp.test.TestClass', 'getWhe_');
descriptions = await index.findDescriptions(criteria);

By default all searches are executed in the context of current classpath that is determined by the current file, unless stated otherwise with an optional parameter. At first index supports only simple queries, but we can add support for fuzzy and regex searches later if required. Index also keeps track of usage and by default returned matches are sorted by score. You can also 'touch' a class to advance its score:

index.touch('com.domain.myapp.test.TestClass');

Actually we probably should use the same indexing mechanism for build properties also. And of course, build properties are fetched in the context of current file by default, just like class descriptions. Therefore the method would become:

const props = await index.findBuildProperties();

It would also provide some helper functions for analyzing current file source. For example sourceHelper.findImportDescriptions() would return all imports in separate 'groups' and additional meta info like buffer positions for those statements. This can be used for folding import statements (currently autocomplete-java does this also).

And of course it would provide some commands that other java packages could also listen:
- Refresh: Refresh class descriptions of current project. Inform listening java packages that they should refresh their caches/indexes/tempdata also.
- Full refresh: Refresh all class descriptions (project classes, libraries, JDK stuff, etc). Inform listening java packages that they should do full refresh also.

And of course some configurations:
- JAVA_HOME
- Build properties (for overriding or for doing without build property files):
  - Classpath
  - Compile target directory
  - Source encoding
  - ...
- For resolving build property file conflicts:
  - Build properties filename: (default: build.properties)
  - Build property load order: (default: properties, eclipse, maven)
- In case of performance issues on slow machines / slow hard drives or other problems:
  - Enable class indexing: true/false
  - Enable class member indexing: true/false
  - Enable source file indexing: true/false
  - Index automatically on project start: true/false
  - Watch file changes: true/false

Fastest way to go would be to encapsulate the current autocomplete-java implementation behind an interface described above. Later I would like to refactor the implementation, since many features were implemented rather quickly and some parts of the implementation should be implemented as separate npm modules. But refactoring can be done later since the current implementation seems to work very well. I've used autocomplete-java in large projects myself and I haven't received any performance or operating system dependent issues either lately.

I hope you don't mind if I create a repository for this, since most of the implementation is coming from autocomplete-java, and I will most likely be the person that does most of the implementation and maintenance. You can surely also contribute if you like. Perhaps the package name should also be changed to something else than atom-javaenv since the scope has changed quite a bit.

**Comment by keskiju** [comment]
Perhaps there should not be a hard dependency between linter-javac and atom-javaenv. Then user would be able to install just linter-javac and define javac arguments in package settings. If, however, user has installed also atom-javaenv, he gets the extra capabilities that it provides for linter-javac: multimodule support (multiple classpaths), build property files in multiple formats, centralized java configuration, commands that affect multiple java modules at the same time.

**Comment by keskiju** [comment]
Another idea. A sourceHelper method that parses a statement in the context of current source file and given buffer position, and returns it as a list of class and class member descriptions.

**Comment by keskiju** [comment]
Basically atom-javaenv is a classpath aware indexing service that indexes java related files by scanning files, watching file changes and parsing files into descriptions. This includes build properties, class files, jar files and java source files. Other java packages can provide additional parsers for the indexer also. Therefore additional build tool support (maven, gradle, etc) can be provided as separate packages.

**Comment by florianb** [comment]
Your proposal sounds good to me. I don't mind if you host the repo, i somehow enjoy maintaining projects but i also support your imagination of the project-structure.

Linter-javac strongly depends on class-loading & config-examination. So i surely will enforce a dependency to that module.

Since i expect you drafting the javaenv (there's no better name yet, eh?) in ES6, i currently see no need for further npm-packages due to the strong module-capablities of ES6.

**Comment by florianb** [comment]
I think adding additional "importers" (f.e. javaenv-maven or whatever) makes only sense, if the user has somehow the control if the stuff gets loaded.

So unless the expected module-size does not extend a arguable limit and we didn't figure out how the user has to react on the need to make use of a third-party-tool config i'd suggest to leave the "plugins" in the project. They should be loaded lazily for sure.

**Comment by keskiju** [comment]
Yeah, I'm ok with that if the plugins are pretty basic. However if tooling support includes lots of build tool specific features, then it should be a separate package.

**Comment by florianb** [comment]
:+1:

I just scratched the surface (while searching for strategies resolving classpath-information), but i expect that using running a command and parsing its output will do the trick in most of the times.

**Comment by keskiju** [comment]
If you think of atom package menu items and commands from the users point of view, "Java Project" might be a good name for the packages.

**Comment by keskiju** [comment]
In eclipse, there is a menu called "Project" and underneath it commands for cleaning and building project.

**Comment by florianb** [comment]
That's a good idea! :100:

**Comment by florianb** [comment]
:thought_balloon: Hmm.. how far will this package go?

**Comment by keskiju** [comment]
Dunno :P I would try to keep the scope defined here. It provides centralized configuration and commands, and indexed info about the java project. All java related fancy UI features should be separate packages. Then users can choose among them, which pleases them most.

**Comment by florianb** [comment]
Yeah, i agree. We probably should summarize which entities of the java-environment might be of interest. Then probably specifying the expected activites (f.e. tracking, resembling by scope, building) could help getting a formal and exhaustive list of topics.

Except from optional debug-messages or warnings i wouldn't implement any ui-features.

**Comment by keskiju** [comment]
Yes, I agree. What's your timetable with linter-javac? I should probably implement at least the findBuildProperties() method with a java property file support before you release linter-javac 2.0.0.

**Comment by florianb** [comment]
I loosely planned the release for April the 1st - it's hopefully becoming a great fun.

But for the 1st release (2.0.0) i only expect the dependency on java_project being working and being able to read the dirty `.classpath`-files, with other words, it should not need to support all the new stuff out of the box.

And do not forget, i am willing to help you.. :)

**Comment by keskiju** [comment]
It'll take a month or so until I have time to start moving implementation from autocomplete-java to atom-java-project. At first I thought I would just move the current .classpath/javaArgs implementation from linter-javac to atom-java-project, encapsulate it behind findBuildProperties() to get you going and later replace it with a real indexed implementation. But that's probably not a good idea, since it would just cause unneccessary changes for users in a short period of time and might lead to confusion:
- First users have to install atom-java-project, which is used only by linter-javac and brings nothing new to the table. They also have to move their existing linter-javac configs to atom-java-project without apparent reason.
- A little later they have to change their .classpath files to another format and make changes to their autocomplete-java configuration also.

If current .classpath functionality suffices, perhaps it's better that you just encapsulate the current .classpath implementation of linter-javac behind a findBuildProperties() method within linter-javac implementation. Then you can easily replace that method call once atom-java-project has been released.

**Comment by keskiju** [comment]
findBuildProperties() would return either a Map or a plain Object. Keys should be prefixed with a tool name, for example:

```
java.home
java.classpath
java.compileTargetDir
java.sourceEncoding
javac.argsfile
ant.home
ant.opts
```

Using a Map:

```
const props = findBuildProperties();
const cp = props.get('java.classpath');
```

Using a plain Object:

```
const props = findBuildProperties();
const cp = props['java.classpath'];
```

Using a hierarchical plain Object:

```
const props = findBuildProperties();
const cp = props['java']['classpath'];
```

Which one would you prefer?

**Comment by florianb** [comment]
Thanks for discussing the stuff with me. :beer: 

At first no worries about the potential hassle for the user-base of `linter-javac`. I'm planning to make the upgrade-procedure completely hassle-free. And i will change the running implementation only if i found a way to change things without disturbing the user.

However - i completely agree with your suggestion (i would prefer if the function would be called `.getBuildProperties`, since i finally expect, that the properties are already found when i try to access them).

I would suggest using a hierarchical Map offering namespaces for every tool (including the meta data of `java-project`):

```
{
  java-project: ...,
  ant: ...,
  maven: ...,
  eclipse: ...,
  gradle: ...
}
```

Unfortunately ES does not provide an Elvis-Operator (aka Safe Navigation Operator like [Rubys `.&`](http://mitrev.net/ruby/2015/11/13/the-operator-in-ruby/)), but if we face deep level operations i'd think about creating a smooth accessor.

**Comment by keskiju** [comment]
Yes I agree. Separate tool namespaces would be useful.

I chose 'find' so that naming would be uniform with other find methods. All of them find project related content from index using the "path of current file" as a context:

```
const props = await index.findBuildProperties();
const classDesc = await index.findClass('com.domain.myapp.test.TestClass');
```

I also try to avoid word 'get' with asynchronous methods because I feel that it has kind of synchronous ring to it (results are returned immediately). Using some other term like 'find', it is more clear that the method should be called using await. All find methods are asynchronous because index may not yet contain information that caller is asking for (e.g. user just opened a new project, executed the 'clean' command or created a new build properties file).

**Comment by florianb** [comment]
I understand - i thought the gathering of information would be done completely independent of the usage. Saying if you call `get` there is no further information gathered and i expected it to by synchronous. This implies that there is a separate mechanism invoking a refresh.

I don't mind your decision to make it asynchronous in general but i'd like to note that i think that your approach may lead to a redundancy since it looks like i have to cache the values to prevent any delay to the execution. I would have expected these values to be stored centralized by the java-project-package.

The java-project-package should also be able to determine if the data has to be refreshed due to changing file-events. A refresh could -- of course -- be enforced by a client-plugin.

**Comment by keskiju** [comment]
You don't have to cache values, since they are already cached in index of java-project package. Most of the time the returned promise is resolved immediately, so usually there is no delay either.

**Comment by keskiju** [comment]
Perhaps atom-java-project should provide also synchronous methods that return null in case of a index miss.

**Comment by florianb** [comment]
Okay - i am excited.. ;)

**Comment by florianb** [comment]
Do suggest introducing `null` to indicate a failure state?

**Comment by keskiju** [comment]
No, just a index miss (requested content is yet to be scanned/indexed, or not found at all). This might be useful if callee wants to perform a operation either immediately and not at all (cancel operation in case of a index miss). Actually most autocomplete features should operate this way. Perhaps asynchronous find methods are not required after all. Synchronous get methods suffice for most use cases and asynchronous find methods can be implemented later if required.

**Comment by florianb** [comment]
Okay, then i'd consider using `Map.size > 0`.

**Comment by keskiju** [comment]
It should work the same way for all methods and not all methods return a map. Undefined or null would work in all cases regardless of return type. Of course throwing an exception is a valid option also.

However getBuildProperties() is somewhat special case, because returned data is gathered both from package settings and project files. Therefore requested data may be partially available. Perhaps it would be simpler just to always use asynchronous findBuildProperties() method for fetching build properties. Then callee can trust that all data is always being returned. There is no advantage in using synchronous get in case of build properties since indexer scans all build properties before any other files so index miss is a really rare case and if index miss occurs, delay would always be really short.

**Comment by florianb** [comment]
Hmm - i wonder if we're able to detect incomplete project-data and to decide properly if a missing information is that critical that f.e. the linter is not working (which is probably only the case if the javac-path is missing).

Probably it's a good idea to somehow track the state of every namespace (uninitialized, ok, faulty)?
